vertx-pojo-mapper-common/src/main/java/de/braintags/vertx/jomnigate/mapping/impl/AbstractMapperFactory.java

```diff
@@ -46,22 +68,180 @@ public AbstractMapperFactory(IDataStore<?, ?> dataStore) {
     return datastore;
   }
 
+  @SuppressWarnings("unchecked")
   @Override
   public final <T> IMapper<T> getMapper(Class<T> mapperClass) {
     String className = mapperClass.getName();
     if (mappedClasses.containsKey(className)) {
-      @SuppressWarnings("unchecked")
-      IMapper<T> cachedEntry = mappedClasses.get(className);
-      return cachedEntry;
+      return (IMapper<T>) mappedClasses.get(className);
     }
     if (!mapperClass.isAnnotationPresent(Entity.class))
       throw new UnsupportedOperationException(String
           .format("The class %s is no mappable entity. Add the annotation Entity to the class", mapperClass.getName()));
+
+    IMapper<T> mapper = createMapperBlocking(mapperClass);
+    Map<String, IMapper<?>> tmpMap = new HashMap<>(mappedClasses);
+    tmpMap.put(className, mapper);
+    mappedClasses = tmpMap;
+    return mapper;
+  }
+
+  private final <T> IMapper<T> createMapperBlocking(Class<T> mapperClass) {
+    IObserverContext context = IObserverContext.createInstance();
+    LOGGER.debug("pre mapping for " + mapperClass.getName());
+    preMapping(mapperClass, context);
+    LOGGER.debug("createMapper for " + mapperClass.getName());
     IMapper<T> mapper = createMapper(mapperClass);
-    mappedClasses.put(className, mapper);
+    LOGGER.debug("post mapping for " + mapperClass.getName());
+    postMapping(mapper, context);
     return mapper;
   }
 
+  private void preMapping(Class<?> mapperClass, IObserverContext context) {
+    CountDownLatch latch = new CountDownLatch(1);
+    ResultObject<Void> ro = new ResultObject<>(null);
+
+    // using only vertx.executeBlocking will potentially block with MongoDatastore
+    // thus - during init - we are using Runnable
+    Runnable runnable = new Runnable() {
+      @Override
+      public void run() {
+        try {
+          LOGGER.debug("start handle before mapping");
+          Future<Void> f = handleBeforeMapping(mapperClass, context);
+          LOGGER.debug("stop handle before mapping");
+          if (f.failed()) {
+            ro.setThrowable(f.cause());
+          } else {
+            // no result to be set;
+          }
+        } catch (Exception e) {
+          ro.setThrowable(e);
+        }
+        latch.countDown();
+      }
+    };
+    Thread thr = new Thread(runnable);
+    thr.start();
```

Comment:
let it for now like that, cause its an internal solution, which is not connected to vertx threads